### DISCUSSION TITLE
alternate embedded:io units for sensors, etc

### DISCUSSION THREAD

**Discussion by tve** [Moddable-OpenSource/moddable#1142] [Q&A]
Per ECMA 419 the ADC returns values in the range 0 .. (1<<resolution)-1, yet I need it in millivolts, and ESP-IDF actually reports in millivolts.
Per ECMA 419 an accelerometer reports values in m/s^2 but I need them in g and the accelerometer I'm using (BMI160) also reports in g.
If I want to "fix these up" is the right process to use the format property? E.g. `adc.format = "millivolt"` and `accel.format = "g"`?

**Comment by phoddie** [comment]
The sensor API provides normalized values. A conforming implementation provides the value in the specified units. The goal is to provide developers with consistency so that they aren't surprised by unit changes when switching components. This matches the approach taken by W3C Sensors.

If the client of the sensor wants different units, it should perform the conversion.

The `format` is not units. It is the [data type](https://419.ecma-international.org/#-9-io-class-pattern-format-property).

**Comment by tve** [comment]
I understand the need for a standard unit. I misread the standard, it says "type of data" not "data type" and in my mind data in g's vs. m/s^2 are two different types...
I find the unit of the ADC particularly annoying because many ADCs do return absolute voltages. I understand not all do and presumably that's why a unitless format, uuh, range was chosen. It means that one has to guess what voltage `(1<<resolution)-1` represents which pretty much defeats the portability anyway.
Oh well...
